ddtrace/contrib/aiokafka/patch.py

```diff
@@ -0,0 +1,175 @@
+
+
+def bootstrap_servers(instance):
+    if hasattr(instance, "client"):
+        client = instance.client
+    if hasattr(instance, "_client"):
+        client = instance._client
+    if client._bootstrap_servers is not None:
+        return ",".join(client._bootstrap_servers)
+
+
+def create_send_span_tags(instance, args, kwargs):
+    topic = get_argument_value(args, kwargs, 0, "topic")
+    tags = {
+        COMPONENT: config.aiokafka.integration_name,
+        SPAN_KIND: SpanKind.PRODUCER,
+        TOPIC: topic,
+        HOST_LIST: bootstrap_servers(instance),
+        MESSAGING_SYSTEM: SERVICE,
+    }
+    partition = get_argument_value(args, kwargs, 3, "partition", True) or None
+    if partition:
+        tags[PARTITION] = partition
+    key = get_argument_value(args, kwargs, 2, "key", True) or None
+    if key:
+        tags[MESSAGE_KEY] = key.decode("utf-8")
+    value = get_argument_value(args, kwargs, 1, "value", True) or None
+    tags[TOMBSTONE] = str(value is None).lower()
+    return tags
+
+
+async def traced_send_and_wait(func, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return await func(*args, **kwargs)
+
+    with core.context_with_data(
+        "aiokafka.send_and_wait",
+        parent=None,
+        span_name=schematize_messaging_operation(PRODUCE, provider="kafka", direction=SpanDirection.OUTBOUND),
+        span_type=SpanTypes.WORKER,
+        service=trace_utils.ext_service(pin, config.aiokafka),
+        tags=create_send_span_tags(instance, args, kwargs),
+        pin=pin,
+    ) as ctx:
+        result = await func(*args, **kwargs)
+        core.dispatch("aiokafka.send_and_wait.post", (ctx,))
+        return result
+
+
+def create_get_span_tags(instance, args, kwargs):
+    tags = {
+        COMPONENT: config.aiokafka.integration_name,
+        SPAN_KIND: SpanKind.CONSUMER,
+        HOST_LIST: bootstrap_servers(instance),
+        MESSAGING_SYSTEM: SERVICE,
+        GROUP_ID: instance._group_id,
+    }
+    return tags
+
+
+async def traced_getone(func, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return await func(*args, **kwargs)
+
+    with core.context_with_data(
+        "aiokafka.getone",
+        parent=None,
+        span_name=schematize_messaging_operation(CONSUME, provider="kafka", direction=SpanDirection.INBOUND),
+        span_type=SpanTypes.WORKER,
+        service=trace_utils.ext_service(pin, config.aiokafka),
+        tags=create_get_span_tags(instance, args, kwargs),
+        pin=pin,
+    ) as ctx:
+        err = None
+        result = None
+        try:
+            result = await func(*args, **kwargs)
+        except Exception as e:
+            err = e
+            raise err
+        finally:
+            _instrument_message(ctx["call"], result, err)
+            ctx["call"].finish()
+        return result
+
+
+def _instrument_message(span, message, err):
```

Comment:
Because it has directly to do with spans, this logic would be more at home in `ddtrace/_trace`, probably as part of an event handler listened for with `core.on`.

Reply:
@emmettbutler I've reworked the code to make use of the event handler listeners.

It does feel like there's no clear split between the integration and tracing code though.

Still to be done is supporting DSM and wrapping other methods for sending/receiving data.

Reply:
DSM is part of https://github.com/DataDog/dd-trace-py/pull/10068/commits/d8774d8a9e5cc1fa9ef1e905c392dfd9d2cd8af9